ts/parsers/paragraphparser.ts

```diff
@@ -1,37 +1,155 @@
+                list.listItems[i]["list"] = this.restructureList(list.listItems[i + 1].list);
+                list.listItems.splice(i + 1, 1);
+            }
+        }
+        return list;
+    }
+
     public static extractParagraphElements(paragraphs: any[]): PowerpointElement["paragraph"] {
         if (!paragraphs || paragraphs.length === 0) {
             return null;
         }
 
-        return paragraphs.map((paragraph) => {
-            const textElements = paragraph["a:r"] || [];
-            const content = textElements.map((txtElement) => {
-                return {
-                    text: txtElement["a:t"] || "",
-                    textCharacterProperties: this.determineTextProperties(getValueAtPath(txtElement, '["a:rPr"][0]'))
-                };
-            });
+        const paras = [];
+        const stack = [];
+        const paragraph: Paragraph = {
+            list: {
+                listType: ListType.Ordered,
+                listItems: []
+            }
+        };
+        let currentParagraph = paragraph;
+        let currentLevel = -1;
 
-            return {
-                content: content,
-                paragraphProperties: this.determineParagraphProperties(paragraph)
-            };
-        });
+        for (const p of paragraphs) {
+            if (this.isList(p)) {
+                const listLevel = this.getListlevel(p);
+
+                //if its the first of the list kind
+                if (currentLevel === -1) {
+                    currentLevel = 0;
+                    currentParagraph.list.listType = this.getListType(p);
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                    stack.push(currentParagraph);
+                }
+                //if the level is same keep pushing the list items in the same array
+                else if (listLevel === currentLevel) {
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                } else if (listLevel > currentLevel) {
+                    //edge case to handle if multiple levels are jumped ahead
+                    //create empty paragraphs/lists to maintain hierarchy and fill in the level gaps
+                    while (currentLevel < listLevel - 1) {
+                        const emptyPara: Paragraph = {
+                            list: {
+                                listType: ListType.UnOrdered,
+                                listItems: []
+                            }
+                        };
+                        currentParagraph.list.listItems.push(emptyPara);
+                        currentParagraph = emptyPara;
+                        //pushing it in the stack to keep track of the parents
+                        stack.push(emptyPara);
+                        currentLevel++;
+                    }
+                    //if there is another hierarchy starting create a new list for it
+                    const newPara: Paragraph = {
```

Comment:
```suggestion
                    const newParagraph: Paragraph = {
```

or you can simply refer to these as 'nodes' if you want to skimp on typing

Reply:
changed it to `newParagraph`